File: src/main/java/ru/arc/hooks/HookRegistry.java
package ru.arc.hooks;

import java.util.HashSet;
import java.util.Set;

import org.bukkit.Bukkit;
import ru.arc.ARC;
import ru.arc.hooks.auraskills.AuraSkillsHook;
import ru.arc.hooks.bank.BankHook;
import ru.arc.hooks.betterstructures.BSListener;
import ru.arc.hooks.citizens.CitizensHook;
import ru.arc.hooks.elitemobs.EMHook;
import ru.arc.hooks.elitemobs.EMListener;
import ru.arc.hooks.lands.LandsHook;
import ru.arc.hooks.lootchest.LootChestHook;
import ru.arc.hooks.luckperms.LuckPermsHook;
import ru.arc.hooks.packetevents.PacketEventsHook;
import ru.arc.hooks.slimefun.BackpackBlockListener;
import ru.arc.hooks.slimefun.SFHook;
import ru.arc.hooks.viaversion.ViaVersionHook;
import ru.arc.hooks.worldguard.WGHook;
import ru.arc.hooks.yamipa.YamipaHook;
import ru.arc.hooks.zauction.AuctionHook;
import ru.arc.hooks.ztranslator.TranslatorHook;
import ru.arc.jobs.JobsModule;
import ru.arc.listeners.BlockListener;
import ru.arc.listeners.CMIListener;
import ru.arc.listeners.ChatListener;
import ru.arc.listeners.CommandListener;
import ru.arc.listeners.IAEvents;
import ru.arc.listeners.JoinListener;
import ru.arc.listeners.PickupListener;
import ru.arc.listeners.RespawnListener;
import ru.arc.listeners.SpawnerListener;

import static org.bukkit.Bukkit.getServer;
import static ru.arc.util.Logging.debug;
import static ru.arc.util.Logging.error;
import static ru.arc.util.Logging.info;

public class HookRegistry {

    public static LandsHook landsHook;
    public static HuskHomesHook huskHomesHook;
    public static PAPIHook papiHook;
    public static CMIHook cmiHook;
    public static ItemsAdderHook itemsAdderHook;
    public static CitizensHook citizensHook;
    public static ViaVersionHook viaVersionHook;
    public static WGHook wgHook;
    public static SFHook sfHook;
    public static EMHook emHook;
    public static YamipaHook yamipaHook;
    public static LuckPermsHook luckPermsHook;
    public static LootChestHook lootChestHook;
    public static AuctionHook auctionHook;
    public static TranslatorHook translatorHook;
    public static boolean jobsEnabled = false;
    public static BankHook bankHook;
    public static RedisEcoHook redisEcoHook;
    public static AuraSkillsHook auraSkillsHook;
    public static PlayerWarpsHook playerWarpsHook;
    public static PacketEventsHook packetEventsHook;
    public static AEHook aeHook;


    public ChatListener chatListener;
    public CommandListener commandListener;
    public SpawnerListener spawnerListener;
    public BlockListener blockListener;
    public JoinListener joinListener;
    public PickupListener pickupListener;
    public IAEvents iaEvents;
    public BetterRTPListener betterRTPListener;
    public RespawnListener respawnListener;
    public BSListener bsListener;
    public ShopListener shopListener;
    public EMListener emListener;

    public void setupHooks() {
        registerVanillaEvents();
        registerHooks();
    }

    public void cancelTasks() {
        if (emHook != null) emHook.cancel();
    }

    private static final Set<String> registeredHooks = new HashSet<>();

    private static void register(String pluginName, boolean single, Runnable runnable) {
        if (registeredHooks.contains(pluginName)) {
            info("Plugin {} already registered", pluginName);
            return;
        }
        if (getServer().getPluginManager().getPlugin(pluginName) != null) {
            info("Registering {} hook", pluginName);
            try {
                runnable.run();
                if (single) {
                    registeredHooks.add(pluginName);
                }
            } catch (Throwable e) {
                error("Error registering {} hook", pluginName, e);
                debug("Hook {} registration failed: {}", pluginName, e.getMessage());
            }
        } else {
            debug("Plugin {} not installed — hook skipped", pluginName);
        }
    }

    /** Registers a hook when any of the given plugin names is present (V3/V4 rename, etc.). */
    private static void registerFirstAvailable(boolean single, Runnable runnable, String... pluginNames) {
        for (String pluginName : pluginNames) {
            if (getServer().getPluginManager().getPlugin(pluginName) != null) {
                register(pluginName, single, runnable);
                return;
            }
        }
        info("Unable to find plugin '{}'", String.join("' or '", pluginNames));
    }

    private void registerHooks() {
        register("PlaceholderAPI", true, () -> {
            papiHook = new PAPIHook();
            papiHook.register();
        });
        register("WorldGuard", true, () -> {
            wgHook = new WGHook();
            Bukkit.getPluginManager().registerEvents(wgHook, ARC.getInstance());
        });

        register("Slimefun", true, () -> {
            sfHook = new SFHook();
            Bukkit.getPluginManager().registerEvents(sfHook, ARC.getInstance());
            Bukkit.getPluginManager().registerEvents(BackpackBlockListener.INSTANCE, ARC.getInstance());
        });

        register("AdvancedEnchantments", true, () -> {
            aeHook = new AEHook();
            Bukkit.getPluginManager().registerEvents(aeHook, ARC.getInstance());
        });

        register("EliteMobs", false, () -> {
            if (emHook == null) emHook = new EMHook();
            emHook.reload();
            if (emListener == null) {
                emListener = new EMListener();
                Bukkit.getPluginManager().registerEvents(emListener, ARC.getInstance());
            }
        });

        register("HuskHomes", true, () -> {
            huskHomesHook = new HuskHomesHook();
            Bukkit.getPluginManager().registerEvents(huskHomesHook, ARC.getInstance());
        });

        register("Lands", true, () -> landsHook = new LandsHook());

        register("Jobs", true, () -> {
            JobsModule.init();
            jobsEnabled = true;
        });

        registerFirstAvailable(true, () -> auctionHook = new AuctionHook(), "zAuctionHouse", "zAuctionHouseV3");

        register("Bank", true, () -> bankHook = new BankHook());

        register("RedisEconomy", true, () -> {
            redisEcoHook = new RedisEcoHook();
            RedisEcoListener redisEcoListener = new RedisEcoListener();
            Bukkit.getPluginManager().registerEvents(redisEcoListener, ARC.getInstance());
        });

        translatorHook = new TranslatorHook();

        register("LuckPerms", true, () -> luckPermsHook = new LuckPermsHook());

        register("AuraSkills", true, () -> auraSkillsHook = new AuraSkillsHook());

        register("CMI", true, () -> {
            cmiHook = new CMIHook();
            CMIListener cmiListener = new CMIListener();
            Bukkit.getPluginManager().registerEvents(cmiListener, ARC.getInstance());
        });

        register("ViaVersion", true, () -> viaVersionHook = new ViaVersionHook());

        register("packetevents", true, () -> packetEventsHook = new PacketEventsHook());

        register("PlayerWarps", true, () -> playerWarpsHook = new PlayerWarpsHook());

        register("LootChest", true, () -> lootChestHook = new LootChestHook());

        register("YamipaPlugin", true, () -> yamipaHook = new YamipaHook());

        register("ItemsAdder", true, () -> {
            itemsAdderHook = new ItemsAdderHook();
            iaEvents = new IAEvents();
            Bukkit.getPluginManager().registerEvents(iaEvents, ARC.getInstance());
        });

        register("Citizens", true, () -> {
            citizensHook = new CitizensHook();
            citizensHook.registerListeners();
        });

        register("BetterRTP", true, () -> {
            betterRTPListener = new BetterRTPListener();
            Bukkit.getPluginManager().registerEvents(betterRTPListener, ARC.getInstance());
        });

        register("BetterStructures", true, () -> {
            bsListener = new BSListener();
            Bukkit.getPluginManager().registerEvents(bsListener, ARC.getInstance());
        });

        register("EconomyShopGUI-Premium", true, () -> {
            shopListener = new ShopListener();
            Bukkit.getPluginManager().registerEvents(shopListener, ARC.getInstance());
        });


    }

    private void registerVanillaEvents() {
        if (chatListener == null) {
            chatListener = new ChatListener();
            Bukkit.getPluginManager().registerEvents(chatListener, ARC.getInstance());
        }

        if (respawnListener == null) {
            respawnListener = new RespawnListener();
            Bukkit.getPluginManager().registerEvents(respawnListener, ARC.getInstance());
        }

        if (spawnerListener == null) {
            spawnerListener = new SpawnerListener();
            Bukkit.getPluginManager().registerEvents(spawnerListener, ARC.getInstance());
        }

        if (joinListener == null) {
            joinListener = new JoinListener();
            Bukkit.getPluginManager().registerEvents(joinListener, ARC.getInstance());
        }

        if (blockListener == null) {
            blockListener = new BlockListener();
            Bukkit.getPluginManager().registerEvents(blockListener, ARC.getInstance());
        }

        if (pickupListener == null) {
            pickupListener = new PickupListener();
            Bukkit.getPluginManager().registerEvents(pickupListener, ARC.getInstance());
        }

        if (commandListener == null) {
            commandListener = new CommandListener();
            Bukkit.getPluginManager().registerEvents(commandListener, ARC.getInstance());
        }
    }

}

File: src/main/kotlin/ru/arc/hooks/HookRegistry.kt
package ru.arc.hooks

import org.bukkit.Bukkit
import org.bukkit.event.HandlerList
import org.bukkit.event.Listener
import ru.arc.ARC
import ru.arc.hooks.auraskills.AuraSkillsHook
import ru.arc.hooks.bank.BankHook
import ru.arc.hooks.betterstructures.BSListener
import ru.arc.hooks.citizens.CitizensHook
import ru.arc.hooks.elitemobs.EMHook
import ru.arc.hooks.elitemobs.EMListener
import ru.arc.hooks.economyshop.EconomyShopGuiPurchaseService
import ru.arc.hooks.economyshop.EconomyShopGuiAuditListener
import ru.arc.hooks.economyshop.ShopPurchaseService
import ru.arc.hooks.lands.LandsHook
import ru.arc.hooks.lootchest.LootChestHook
import ru.arc.hooks.luckperms.LuckPermsHook
import ru.arc.hooks.packetevents.PacketEventsHook
import ru.arc.hooks.slimefun.BackpackBlockListener
import ru.arc.hooks.slimefun.SFHook
import ru.arc.hooks.viaversion.ViaVersionHook
import ru.arc.hooks.worldguard.WGHook
import ru.arc.hooks.yamipa.YamipaHook
import ru.arc.hooks.zauction.AuctionHook
import ru.arc.hooks.zauction.AuctionTrophyGuardListener
import ru.arc.hooks.zauction.AuctionSaleNotifier
import ru.arc.hooks.ztranslator.TranslatorHook
import ru.arc.jobs.JobsModule
import ru.arc.listeners.BlockListener
import ru.arc.listeners.CMIListener
import ru.arc.listeners.ChatListener
import ru.arc.listeners.CommandListener
import ru.arc.listeners.JoinListener
import ru.arc.listeners.PickupListener
import ru.arc.listeners.RespawnListener
import ru.arc.listeners.SpawnerListener
import ru.arc.contracts.SeasonTrophyProtectionListener
import ru.arc.util.Logging.debug
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info

class HookRegistry(
    private val stopJobs: () -> Unit = JobsModule::shutdown,
) : AutoCloseable {
    var chatListener: ChatListener? = null
    var commandListener: CommandListener? = null
    var spawnerListener: SpawnerListener? = null
    var blockListener: BlockListener? = null
    var joinListener: JoinListener? = null
    var pickupListener: PickupListener? = null
    var betterRTPListener: BetterRTPListener? = null
    var leafRTPListener: LeafRTPListener? = null
    var respawnListener: RespawnListener? = null
    var bsListener: BSListener? = null
    var emListener: EMListener? = null
    var seasonTrophyProtectionListener: SeasonTrophyProtectionListener? = null

    private val registeredHooks = HashSet<String>()
    private val registeredListeners = LinkedHashSet<Listener>()

    internal var isClosed: Boolean = false
        private set

    companion object {
        @JvmField var landsHook: LandsHook? = null

        @JvmField var huskHomesHook: HuskHomesHook? = null

        @JvmField var papiHook: PAPIHook? = null

        @JvmField var cmiHook: CMIHook? = null

        @JvmField var itemsAdderHook: ItemsAdderHook? = null

        @JvmField var citizensHook: CitizensHook? = null

        @JvmField var viaVersionHook: ViaVersionHook? = null

        @JvmField var wgHook: WGHook? = null

        @JvmField var sfHook: SFHook? = null

        @JvmField var emHook: EMHook? = null

        @JvmField var yamipaHook: YamipaHook? = null

        @JvmField var luckPermsHook: LuckPermsHook? = null

        @JvmField var lootChestHook: LootChestHook? = null

        @JvmField internal var auctionHook: AuctionHook? = null

        @JvmField var translatorHook: TranslatorHook? = null

        @JvmField var jobsEnabled: Boolean = false

        @JvmField var bankHook: BankHook? = null

        @JvmField var redisEcoHook: RedisEcoHook? = null

        @JvmField var auraSkillsHook: AuraSkillsHook? = null

        @JvmField var playerWarpsHook: PlayerWarpsHook? = null

        @JvmField var packetEventsHook: PacketEventsHook? = null

        @JvmField var aeHook: AEHook? = null

        @JvmField var myWorldsHook: MyWorldsHook? = null

        @JvmField var partiesHook: PartiesHook? = null

        internal var shopPurchaseService: ShopPurchaseService? = null

        private fun clearGlobalHooks() {
            landsHook = null
            huskHomesHook = null
            papiHook = null
            cmiHook = null
            itemsAdderHook = null
            citizensHook = null
            viaVersionHook = null
            wgHook = null
            sfHook = null
            emHook = null
            yamipaHook = null
            luckPermsHook = null
            lootChestHook = null
            auctionHook = null
            translatorHook = null
            jobsEnabled = false
            bankHook = null
            redisEcoHook = null
            auraSkillsHook = null
            playerWarpsHook = null
            packetEventsHook = null
            aeHook = null
            myWorldsHook = null
            partiesHook = null
            shopPurchaseService = null
        }
    }

    fun setupHooks() {
        check(!isClosed) { "HookRegistry is closed" }
        registerVanillaEvents()
        registerHooks()
    }

    @Deprecated("Use close()", ReplaceWith("close()"))
    fun cancelTasks() {
        close()
    }

    override fun close() {
        if (isClosed) return
        isClosed = true

        val failures = mutableListOf<Throwable>()
        cleanup(failures) { emHook?.close() }
        cleanup(failures) { auctionHook?.close() }
        cleanup(failures) { citizensHook?.close() }
        cleanup(failures) { papiHook?.unregister() }
        if (jobsEnabled) {
            cleanup(failures, stopJobs)
        }
        jobsEnabled = false

        registeredListeners.forEach { listener ->
            cleanup(failures) { HandlerList.unregisterAll(listener) }
        }
        registeredListeners.clear()
        registeredHooks.clear()
        clearInstanceListeners()
        clearGlobalHooks()

        if (failures.isNotEmpty()) {
            val first = failures.first()
            failures.drop(1).forEach(first::addSuppressed)
            throw first
        }
    }

    private fun cleanup(
        failures: MutableList<Throwable>,
        action: () -> Unit,
    ) {
        try {
            action()
        } catch (failure: Throwable) {
            failures += failure
        }
    }

    private fun clearInstanceListeners() {
        chatListener = null
        commandListener = null
        spawnerListener = null
        blockListener = null
        joinListener = null
        pickupListener = null
        betterRTPListener = null
        leafRTPListener = null
        respawnListener = null
        bsListener = null
        emListener = null
        seasonTrophyProtectionListener = null
    }

    private fun register(
        pluginName: String,
        single: Boolean,
        action: () -> Unit,
    ) {
        if (pluginName in registeredHooks) {
            info("Plugin {} already registered", pluginName)
            return
        }
        if (Bukkit.getServer().pluginManager.getPlugin(pluginName) == null) {
            debug("Plugin {} not installed — hook skipped", pluginName)
            return
        }

        info("Registering {} hook", pluginName)
        val listenersBefore = registeredListeners.toSet()
        try {
            action()
            if (single) registeredHooks += pluginName
        } catch (failure: Throwable) {
            val partiallyRegistered = registeredListeners - listenersBefore
            partiallyRegistered.forEach { listener ->
                runCatching { HandlerList.unregisterAll(listener) }
            }
            registeredListeners.removeAll(partiallyRegistered.toSet())
            error("Error registering {} hook", pluginName, failure)
            debug("Hook {} registration failed: {}", pluginName, failure.message)
        }
    }

    private fun registerFirstAvailable(
        single: Boolean,
        action: () -> Unit,
        vararg pluginNames: String,
    ) {
        for (pluginName in pluginNames) {
            if (Bukkit.getServer().pluginManager.getPlugin(pluginName) != null) {
                register(pluginName, single, action)
                return
            }
        }
        info("Unable to find plugin '{}'", pluginNames.joinToString("' or '"))
    }

    private fun <T : Listener> registerListener(listener: T): T {
        Bukkit.getPluginManager().registerEvents(listener, ARC.instance)
        registeredListeners += listener
        return listener
    }

    private fun registerHooks() {
        register("PlaceholderAPI", true) {
            val hook = PAPIHook()
            check(hook.register()) { "PlaceholderAPI rejected ARC expansion registration" }
            papiHook = hook
        }
        register("WorldGuard", true) {
            wgHook = registerListener(WGHook())
        }
        register("Slimefun", true) {
            val hook = registerListener(SFHook())
            registerListener(BackpackBlockListener)
            sfHook = hook
            runCatching { hook.registerOptionalDenizenTags() }
                .onSuccess { tags ->
                    if (tags.isNotEmpty()) {
                        info("Registered optional Denizen Slimefun tags: {}", tags.joinToString())
                    }
                }
                .onFailure { failure ->
                    error("Unable to register optional Denizen Slimefun tags", failure)
                }
        }
        register("AdvancedEnchantments", true) {
            aeHook = registerListener(AEHook())
        }
        register("EliteMobs", false) {
            val existingHook = emHook
            val hook = existingHook ?: EMHook()
            try {
                hook.reload()
                val listener = emListener ?: registerListener(EMListener())
                emHook = hook
                emListener = listener
            } catch (failure: Throwable) {
                if (existingHook == null) hook.close()
                throw failure
            }
        }
        register("HuskHomes", true) {
            huskHomesHook = registerListener(HuskHomesHook())
        }
        register("Lands", true) { landsHook = LandsHook() }
        register("Jobs", false) {
            if (!jobsEnabled) jobsEnabled = JobsModule.init()
        }
        registerFirstAvailable(
            true,
            {
                val hook = AuctionHook()
                try {
                    hook.start()
                    registerListener(AuctionTrophyGuardListener())
                    registerListener(AuctionSaleNotifier(hook))
                    auctionHook = hook
                } catch (failure: Throwable) {
                    hook.close()
                    throw failure
                }
            },
            "zAuctionHouse",
            "zAuctionHouseV3",
        )
        register("Bank", true) { bankHook = BankHook() }
        register("Parties", true) { partiesHook = PartiesHook() }
        register("RedisEconomy", true) {
            val hook = RedisEcoHook()
            registerListener(RedisEcoListener())
            redisEcoHook = hook
        }
        if (translatorHook == null) translatorHook = TranslatorHook()
        register("LuckPerms", true) { luckPermsHook = LuckPermsHook() }
        register("AuraSkills", true) { auraSkillsHook = AuraSkillsHook() }
        register("CMI", true) {
            val hook = CMIHook()
            registerListener(CMIListener())
            cmiHook = hook
        }
        register("ViaVersion", true) { viaVersionHook = ViaVersionHook() }
        register("packetevents", true) { packetEventsHook = PacketEventsHook() }
        register("PlayerWarps", true) { playerWarpsHook = PlayerWarpsHook() }
        register("LootChest", true) { lootChestHook = LootChestHook() }
        register("YamipaPlugin", true) { yamipaHook = YamipaHook() }
        register("ItemsAdder", true) {
            val itemsAdder = checkNotNull(Bukkit.getPluginManager().getPlugin("ItemsAdder"))
            val resourcePackZip = itemsAdder.dataFolder.toPath().resolve("output/generated.zip")
            val syncScript = BundledResourcePackSyncScript.install(ARC.instance.dataFolder.toPath())
            val config = ResourcePackSyncConfig.load(ARC.instance.dataFolder.toPath())
            itemsAdderHook = registerListener(ItemsAdderHook(resourcePackZip, syncScript, config))
        }
        register("Citizens", true) {
            citizensHook = CitizensHook()
        }
        register("BetterRTP", true) {
            betterRTPListener = registerListener(BetterRTPListener())
        }
        register("RTP", true) {
            leafRTPListener = registerListener(LeafRTPListener())
        }
        register("My_Worlds", true) {
            val plugin = checkNotNull(Bukkit.getPluginManager().getPlugin("My_Worlds"))
            myWorldsHook = MyWorldsHook(plugin)
        }
        register("BetterStructures", true) {
            bsListener = registerListener(BSListener())
        }
        register("EconomyShopGUI-Premium", true) {
            val translator = checkNotNull(translatorHook) { "Material translator is not initialized" }
            val purchaseService = EconomyShopGuiPurchaseService { item -> translator.translate(item) }
            registerListener(EconomyShopGuiAuditListener())
            shopPurchaseService = purchaseService
        }
    }

    private fun registerVanillaEvents() {
        if (chatListener == null) {
            chatListener = registerListener(ChatListener())
        }
        if (respawnListener == null) {
            respawnListener = registerListener(RespawnListener())
        }
        if (spawnerListener == null) {
            spawnerListener = registerListener(SpawnerListener())
        }
        if (joinListener == null) {
            joinListener = registerListener(JoinListener())
        }
        if (blockListener == null) {
            blockListener = registerListener(BlockListener())
        }
        if (pickupListener == null) {
            pickupListener = registerListener(PickupListener())
        }
        if (commandListener == null) {
            commandListener = registerListener(CommandListener())
        }
        if (seasonTrophyProtectionListener == null) {
            seasonTrophyProtectionListener = registerListener(SeasonTrophyProtectionListener())
        }
    }
}
